crates/primitives/src/utilities.rs

```diff
@@ -31,6 +33,56 @@ pub fn create2_address(caller: B160, code_hash: B256, salt: U256) -> B160 {
     B160(hasher.finalize().as_slice()[12..].try_into().unwrap())
 }
 
+/// Calculates the `excess_blob_gas` from the parent header's `blob_gas_used` and `excess_blob_gas`.
+///
+/// See also [the EIP-4844 helpers](https://eips.ethereum.org/EIPS/eip-4844#helpers).
+#[inline]
+pub fn calc_excess_blob_gas(parent_excess_blob_gas: u64, parent_blob_gas_used: u64) -> u64 {
+    (parent_excess_blob_gas + parent_blob_gas_used).saturating_sub(TARGET_BLOB_GAS_PER_BLOCK)
+}
+
+/// Calculates the blobfee from the header's excess blob gas field.
+///
+/// See also [the EIP-4844 helpers](https://eips.ethereum.org/EIPS/eip-4844#helpers).
+#[inline]
+pub fn calc_blob_fee(excess_blob_gas: u64) -> u64 {
+    fake_exponential(
+        MIN_BLOB_GASPRICE,
+        excess_blob_gas,
+        BLOB_GASPRICE_UPDATE_FRACTION,
+    )
+}
+
+/// Approximates `factor * e ** (numerator / denominator)` using Taylor expansion.
+///
+/// This is used to calculate the blob price.
+///
+/// See also [the EIP-4844 helpers](https://eips.ethereum.org/EIPS/eip-4844#helpers).
+#[inline]
+pub fn fake_exponential(factor: u64, numerator: u64, denominator: u64) -> u64 {
+    assert!(denominator > 0, "attempt to divide by zero");
+    let factor = factor as u128;
+    let numerator = numerator as u128;
+    let denominator = denominator as u128;
+
+    let mut i = 1;
+    let mut output = 0;
+    let mut numerator_accum = factor * denominator;
+    while numerator_accum > 0 {
+        output += numerator_accum;
+
+        // SAFETY: asserted > 0 above.
+        // This is needed to eliminate the compiler's divide by 0 check.
+        numerator_accum = unsafe {
+            (numerator_accum * numerator)
+                .checked_div(denominator * i)
+                .unwrap_unchecked()
```

Comment:
Is it better to remove unsafe and use `unwrap_or_default()`?

Reply:
it can only divide by zero if the `denominator` arg is 0 (`i` starts at 1), which is why I added the assertion above

Reply:
with `unwrap_or_default` would remove need of unsafe, or maybe we can just use `.expect("Denominator is checked above")`

Reply:
The problem is that the compiler does not optimize the check out, when we know for a fact that this cannot overflow.

Reply:
You are thinking too deeply about optimization. As this function has a specific purpose that we know we can make an assumption that this is not going to be a hot loop that we expect to be called a few billion times in the loop, so it is okay to have it be slightly inefficient for more readable code.

Reply:
defer to Dragan here - agree we don't need to be "too smart" about it given it's only called once per block?

cc @mattsse,